### ISSUE TITLE
write_back does not work for me, unclear about its usage

### ISSUE THREAD

**Issue by bluppfisk** [msiemens/tinydb#204]
I have a database (strongly simplified):

```
{"_default": {"1": {"name": "Blah", "location": ["Belgium"]}, "2": {"name": "Blah", "location": ["France"]}}}
```

I want to replace each of these locations with a different value, so following the [documentation](http://tinydb.readthedocs.io/en/latest/usage.html), this should be possible by doing the following:

```
	def test():
		items = db.all()
		for item in items:
			new_location = "Germany"
			item.location = new_location

			print(new_location)
			print(item.location)

		print(items)
		db.write_back(items)
```

Actual output:

```
Germany
Germany
Germany
Germany
[{'location': ['Belgium'], 'name': 'Blah'}, {'location': ['France'], 'name': 'Blah'}]
```

Expected output for items:
```
[{'location': ['Germany'], 'name': 'Blah'}, {'location': ['Germany'], 'name': 'Blah'}]
```

So somehow it is not manipulating the document correctly, but only the object that holds the document, which seems to be a different thing. I'm confused about this as I feel I'm following the documentation example pretty closely (except I use `all()` instead of `search()`), but both return a list of documents with their associated doc_ids so I don't see how that would be different.

**Comment by Phuket2** [comment]
@bluppfisk, I know you have cut down the code you are showing here.  The below test code works for me.  It appears that the write_back method is popping items out of the list that is passed to it as its writing them back to the database. I didn't look at the source code, but I guess thats what is happening. 
Anyway, hope it helps.  BTW, I am using tinydb version 3.8.0

```python
try:
    from tinydb import TinyDB, __version__ as tinydb_version
    print('tinydb version {}\n'.format(tinydb_version))
except:
    print('This may not work as expected. The version of tinydb may need updating')

import copy
from faker import Faker
fake = Faker()

def make_dummy_data(num_items=100):
    '''
    Create some dummy people to use...
    returns a list of PersonClass objects 
    built up using Faker
    '''
    return [
                {'idx': i,
                'first': fake.first_name(),
                'last': fake.last_name(),
                'age': fake.random_int(min=18, max=106),
                'some_date':fake.date(pattern="%Y-%m-%d"),
                }
                for i in range(0, num_items)
            ]
            
db_fn = 'wb_test.json'
some_data = make_dummy_data(10)        # add some test data
db = TinyDB(db_fn)                     
db.purge_tables()                      # for testing clear the tables
db.insert_multiple(some_data)          # insert the test data into the db

docs = db.all()                        # a reference to a list from db.all()
                                       # I checked the type to be sure it was
                                       # not a generator etc...
for doc in docs:
    doc['age'] = 17                    # set all the ages to 17. for the test
                                       # in the test data, there will be no ages
                                       # less than 18.

db.write_back(docs)                    # write the changes back

for doc in db.all():                   
    print(doc)                         # re print the docs in the db to
                                       # make sure they were updated
```

**Comment by bluppfisk** [comment]
@Phuket2 , so, if I modify the code to read `item['location'] = new_location` instead of `item.location = new_location`, it works. However, the documentation reads that you can use the dot notation, which seems to be incorrect.

**Comment by Phuket2** [comment]
@bluppfisk , yeah the docs seem to be wrong on that one. I just looked at this the other day, I replied to another post here about the dot notation  [link](https://github.com/msiemens/tinydb/issues/199) The Document object is defined in the database.py source file. The Document class subclasses dict, so its basically a dict.  But it has an additional instance attr called doc_id.  The doc_id can be accessed with the dot notation, but the data inside has to be accessed as you access data in a dict i.e with the key notation. Sorry, if my explaination is not clear.
